# MScA, Advanced Machine Learning (32009)

# Week 2 Workshop

# Programming with RDD

### Yuri Balasanov, &copy; iLykei, 2017

Main text: Jeffrey Aven, Sams Teach Yourself Apache SPARK in 24 Hours,Pearson Education, Inc., 2017 

Fundamental object in Spark programming is called Resilient Distributed Dataset (**RDD**). <br>

-  RDDs are resilient because if a node performing a Spark operation failes the dataset can be reconstructed. This is achieved by Spark knows the lineage of precursors of each RDD
-  RDDs are distributed because data are partitionedand distributed as in-memory collections of objects across workers in the cluster
-  RDDs are datasets consisting of records, each of which uniquely identifiable. Records can be collections of fields, like rows in a table of a relational database, or lines of text in a file, or some other format.

RDDs are **immutable**: after they are created they cannot be updated. Instead of updating an RDD a new RDD will be created with lineage tracked. Child RDDs can be created by by performing a transformation, like Map or Filter functions or performing an action like, for example count. 

## Loading data in RDD

### Loading data from file or files

There are 2 main finctions for creating RDDs from text files:

-  `sc.textFile(name,minPartitions=None,use_unicode=True)` - importing a single file; each line of the file represents a record
-  `sc.wholeTextFiles()` importing a group or directory of files each of which becomes a record the file name as key and file content as value

<font color=green>**Example**</font> <br>

<font color=green>Read one of the license files for Spark</font>


In [1]:
lic=sc.textFile("C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/LICENSE-py4j.txt")
print(lic.take(100))

['Copyright (c) 2009-2011, Barthelemy Dagenais All rights reserved.', '', 'Redistribution and use in source and binary forms, with or without', 'modification, are permitted provided that the following conditions are met:', '', '- Redistributions of source code must retain the above copyright notice, this', 'list of conditions and the following disclaimer.', '', '- Redistributions in binary form must reproduce the above copyright notice,', 'this list of conditions and the following disclaimer in the documentation', 'and/or other materials provided with the distribution.', '', '- The name of the author may not be used to endorse or promote products', 'derived from this software without specific prior written permission.', '', 'THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"', 'AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE', 'IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE', 'ARE DISCLAIMED. IN NO EVENT SHA

<font color=green>Read one of the license files for SparTotal number of lines in the file:</font>


In [2]:
lic.getNumPartitions()

2

In [3]:
lic.count()

27

<font color=green>Read one of the license files for Spark</font>
Read all files from the licence folder

In [4]:
licAll=sc.textFile("C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/*.txt")
licAll.take(2)

['The MIT License (MIT)', '']

In [6]:
licAll.getNumPartitions()

36

<font color=green>This means that each file is a partition.</font>

<font color=green>Count total number of lines in all files:</font>

In [5]:
licAll.count()

1075

<font color=green>Use `wholeTextFiles()` instead.</font>


In [6]:
licWhole=sc.wholeTextFiles("C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/*.txt")
licWhole.take(2)

[('file:/C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/LICENSE-AnchorJS.txt',
  'The MIT License (MIT)\n\nCopyright (c) <year> <copyright holders>\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the "Software"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in\nall copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS 

In [7]:
licWhole.getNumPartitions()

2

<font color=green>Now the entire folder is one file with 2 partitions, each file is a record:</font>

In [8]:
licWhole.count()

36

In [9]:
licWholeNames=licWhole.keys()
licWholeNames.collect()[1:5]

['file:/C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/LICENSE-antlr.txt',
 'file:/C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/LICENSE-boto.txt',
 'file:/C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/LICENSE-cloudpickle.txt',
 'file:/C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/LICENSE-d3.min.js.txt']

In [10]:
licWhole.values().take(1)

['The MIT License (MIT)\n\nCopyright (c) <year> <copyright holders>\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the "Software"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in\nall copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHER

### Creating RDD programmatically

RDDs can be created by methods like, for example, `sc.parallelize()` or `sc.range()`.

#### Parallelize 

Method `parallelize()` has syntax <br>

`sc.parallelize(c, numSlices=None)`, <br>

where `c` is a collection - a list 

<font color=green>
**Example**
</font>

In [11]:
parallelrdd=sc.parallelize(range(10))
parallelrdd

PythonRDD[16] at RDD at PythonRDD.scala:48

In [12]:
parallelrdd.min()

0

In [13]:
parallelrdd.max()

9

In [14]:
parallelrdd.collect()

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

#### Range 

Method `range` has syntax <br>

`sc.range(start,end=1,step=1,numSlices=None)`.

Create RDD with 1000 integers starting from 0 incrementing by 1:

<font color=green>**Example**</font>

In [15]:
rangerdd=sc.range(0,1000,1,2)
rangerdd

PythonRDD[20] at RDD at PythonRDD.scala:48

In [16]:
rangerdd.getNumPartitions()

2

In [17]:
rangerdd.min()

0

In [18]:
rangerdd.max()

999

In [19]:
rangerdd.take(5)

[0, 1, 2, 3, 4]

## Operations on RDD: Transformations

There are 2 main classes of operations on RDD: Transformations and Actions.

Transformation is an operation applied to each element of RDD which results in new RDD.
Most common transformations are mapping and filtering functions, among them are `map()` and `filter()`.

### Functional Transformations

#### Map()

Syntax:

`.map(<function>, preservePartitioning=False)`

<font color=green>**Example**</font> <br>
<font color=green>$x$ below is a predictor for linear regression model.</font>
<font color=green>Prepare predictors for polynomial regression: $x,~x^2$</font> 

In [20]:
from numpy.random import permutation
x=sc.parallelize(permutation(range(10)))
print(x.collect())
xPolynomial=x.map(lambda element: (element,element**2))
xPolynomial.collect()

[9, 5, 4, 8, 6, 1, 7, 2, 3, 0]


[(9, 81),
 (5, 25),
 (4, 16),
 (8, 64),
 (6, 36),
 (1, 1),
 (7, 49),
 (2, 4),
 (3, 9),
 (0, 0)]

#### FlatMap()

Syntax:

`.flatMap(<function>, preservePartitioning=False)`

Instead of creating a list from each record of the original RDD this function creates a single list, "flattening" the nested structure.

<font color=green>**Example**</font>

In [21]:
xPolynomialFlat=x.flatMap(lambda element: (element,element**2))
xPolynomialFlat.collect()

[9, 81, 5, 25, 4, 16, 8, 64, 6, 36, 1, 1, 7, 49, 2, 4, 3, 9, 0, 0]

#### Filter()

Syntax:

`.filter(<function>)`

The `filter` transformation evaluates a Boolean expression of each element (record) of the original RDD. The returned Boolean value determines whether the corresponding element is returned in the output RDD.


<font color=green>**Example**</font> <br>
<br>
<font color=green>Filter OUT even numbers from sequence of integers 0 to 8.</font>

In [22]:
originalrdd=sc.parallelize([0,1,2,3,4,5,6,7,8])
newrdd=originalrdd.filter(lambda x: x % 2)
newrdd.collect()

[1, 3, 5, 7]

Principle of big data programming: <br>
**FILTER EARLY, FILTER OFTEN**. <br>
It helps avoiding carrying unnecessary records through a process.

<font color=red>**Exercise. Transform the text**</font> <br>

<font color=blue>Create an RDD `txtDoc`</font>

In [23]:
txtDoc=sc.parallelize(['These violent delights have violent ends',
                       'And in their triump die like fire and powder',
                       'Which as they kiss consume'])
txtDocUpper=txtDoc.map(lambda x: x.upper()).flatMap(lambda x: x.split(" ")).filter(lambda x: len(x)>4)

txtDocUpper.collect()

['THESE',
 'VIOLENT',
 'DELIGHTS',
 'VIOLENT',
 'THEIR',
 'TRIUMP',
 'POWDER',
 'WHICH',
 'CONSUME']

<font color=blue>
-  Convert `txtDoc` to upper case to make it look like this <br>
<br>
['THESE VIOLENT DELIGHTS HAVE VIOLENT ENDS', <br>
 'AND IN THEIR TRIUMP DIE LIKE FIRE AND POWDER', <br>
 'WHICH AS THEY KISS CONSUME'] <br>
<br> 
-  Split the text into a combined list of separate words <br>
<br>
['THESE', 'VIOLENT', 'DELIGHTS', 'HAVE', 'VIOLENT', 'ENDS', 'AND', 'IN', 'THEIR', 'TRIUMP', 'DIE', 'LIKE', 'FIRE', 'AND', 'POWDER', 'WHICH', 'AS', 'THEY', 'KISS', 'CONSUME'] <br>
<br> 
-  Select only words that are longer than 4 letters <br>
<br>
['THESE',
 'VIOLENT',
 'DELIGHTS',
 'VIOLENT',
 'THEIR',
 'TRIUMP',
 'POWDER',
 'WHICH',
 'CONSUME']
<br> 
 Enter code in the following cell.
 </font>

In [25]:
#Skipped code



### Grouping, Sorting, Elimination of Duplicates Transformations

The following 3 transformations are for grouping, sorting and eliminating duplicates.

#### GroupBy()

Syntax:

`.groupBy(<function>, numPartitions=None)`

This transformation returns an RDD of items  grouped by a given function.
The function may either specify a particular key by which to group or specify an expression to be evaluated with elements to specify the group (for example, group odd and even numbers separately).

**This functions is not recommended for aggregation of values (like sum or count). The  `groupBy` transformation does not do any aggregation before shuffling data, it also requires that all values for a given key fit into memory**.

#### SortBy()

Syntax is:

`.sortBy(<keyfunc>, ascending=True, numPartitions=None)`

This transformation sorts an RDD by the function that defines the key for a given dataset.

#### Distinct()

Syntax is:

`.distinct(numPartitions=None)`

This transformation returns a new RDD containing distinct elements. It is used to remove duplicates.

<font color=red>**Exercise. Select long words and sort them**</font> <br>
<font color=blue>
<br>
First 3 steps below repeat the steps of the previous exercise.   <br>
<br>
1 Turn lines from Shakespeare in `txtDoc` created earlier into upper case <br>
<br>
['THESE VIOLENT DELIGHTS HAVE VIOLENT ENDS', <br>
 'AND IN THEIR TRIUMP DIE LIKE FIRE AND POWDER', <br>
 'WHICH AS THEY KISS CONSUME'] <br>
 <br>
2 Split the RDD by space <br>
<br>
['THESE', 'VIOLENT', 'DELIGHTS', 'HAVE', 'VIOLENT', 'ENDS', 'AND', 'IN', 'THEIR', 'TRIUMP', 'DIE', 'LIKE', 'FIRE', 'AND', 'POWDER', 'WHICH', 'AS', 'THEY', 'KISS', 'CONSUME'] <br>
<br>
3 Select words longer than 4 letters <br>
<br>
['THESE', 'VIOLENT', 'DELIGHTS', 'VIOLENT', 'THEIR', 'TRIUMP', 'POWDER', 'WHICH', 'CONSUME'] <br>
<br>
4 Select distinct words <br>
<br>
['VIOLENT', 'THEIR', 'THESE', 'TRIUMP', 'POWDER', 'DELIGHTS', 'WHICH', 'CONSUME'] <br>
<br>
5 Sort distinct words in alphabetic order <br>
<br>
['CONSUME', 'DELIGHTS', 'POWDER', 'THEIR', 'THESE', 'TRIUMP', 'VIOLENT', 'WHICH'] <br>
 </font>

In [24]:
#Skipped code
txtDocUpperAlpha= txtDocUpper.distinct().sortBy(lambda x: x)
txtDocUpperAlpha.collect()

['CONSUME',
 'DELIGHTS',
 'POWDER',
 'THEIR',
 'THESE',
 'TRIUMP',
 'VIOLENT',
 'WHICH']

### Set Transformations

The following 3 transformations are set operations

#### Union()

Syntax is:

`RDD1.union(RDD2)`

This function outputs an RDD which is `RDD2` appended to `RDD1`. 
The two RDDs are not required to have the same structure.
Duplicates from the two RDDs are not filtered out.
The output RDD is not sorted.



<font color=green>**Example** <br>
<br>
Create another RDD with Sheakespear's lines and union it with `txtDoc`
</font>

In [34]:
txtDoc2=sc.parallelize(['My bounty is as boundless as the sea',
                        'My love as deep the more I give to thee',
                        'The more I have for both are infinite'])
txtDocUnion = txtDoc.union(txtDoc2)
txtDocUnion.collect()

['These violent delights have violent ends',
 'And in their triump die like fire and powder',
 'Which as they kiss consume',
 'My bounty is as boundless as the sea',
 'My love as deep the more I give to thee',
 'The more I have for both are infinite']

#### Intersection()

Syntax is:

`RDD1.intersection(RDD2)`

This function returns elements that are present in both RDDs.

<font color=red>**Exercise. Intersect documents**</font> <br>
<font color=blue>
<br>
Create `txtDoc3` with more lines from Sheakspear and intersect it with `txtDoc2`. <br>
<br>
 </font>

In [33]:
txtDoc3=sc.parallelize(['Do not swear by the moon', 
                        'For she changes constantly', 
                        'Then your love would also change'])
txtDocInt=txtDoc2.intersection(txtDoc3)
txtDocInt.collect()


[]

<font color=blue>
1 Split `txtDoc3` into words by space <br>
2 Intersect `txtDoc2` with `txtDoc3` <br>
3 Turn the intersection into upper case and print it: <br>
<br>
['THE', 'LOVE']
</font>

In [41]:
#Skipped code
txtDoc3flat = txtDoc3.flatMap(lambda x: x.split(" "))
txtDoc2flat = txtDoc2.flatMap(lambda x: x.split(" "))
txtDocInt2=txtDoc2flat.intersection(txtDoc3flat)
txtDocIntUpper = txtDocInt2.map(lambda x: x.upper())
txtDocIntUpper.collect()

['THE', 'LOVE']

#### Subtract()

Syntax is:

`RDD1.subtract(RDD2,numPartitions=None)`

This transformation returns an RDD that contains all elements of `RDD1` that are not present in `RDD2`.

<font color=green>
**Example** <br>
<br>
Subtract sequence of Fibonacci numbers from sequence of odd numbers.
</font>

In [42]:
odds=sc.parallelize([1, 3, 5, 7, 9])
fibonacci=sc.parallelize([0, 1, 2, 3, 5, 8])
odds.subtract(fibonacci).collect()

[7, 9]

### Sampling 

Before applying any machine learning methods to data in Spark it is very important to learn about the data as much as possible, make cleaning and necessary transformations and preparing the data for machine learning methods.

One of the challenges of big distributed data is that data cannot be as easily explored as in the "spreadsheet" mode.

It may be useful to work first with a sample from data. There are several sampling transformations provided by Spark.

#### Sample()

Syntax of `sample()` is

`.sample(withReplacement, fraction, seed=None)`

It is used to create a random subset of the RDD.
The arguments are:
-  `withReplacement`: Boolean value specifying whether elements of the RDD are drawn with replacement or not;
-  `fraction`: a double value between 0 and 1 representing probability of drawing element from the RDD; effectively this argument sets the size of the sample;
-  `seed`: optional seed for random number generator.

<font color=green>
**Example** <br>
<br>
Create an RDD from a license file. <br>
Count number of records. <br>
Sample records with probability 0.1
</font>

In [43]:
licAll=sc.textFile("C:/dev/spark-2.2.0-bin-hadoop2.7/spark-2.2.0-bin-hadoop2.7/licenses/*.txt")
print(licAll.count())
sampled_licAll=licAll.sample(False,0.1,seed=5)
print(sampled_licAll.count())

1075
100


<font color=green>
Returned sampled RDD is approximately 10% of the original RDD.
</font>



## Actions

Action is an operation on RDD which returnes values or data to the Driver process.
They are typically the final step of a Spark program.
Most common actions include `reduce()`, `collect()`, `count()` and `saveAsTextFile()`.
The following action returns to Driver the content od `newrdd`.

#### Count()

Syntax is:

`.count()`

This function counts elements.

<font color=green>
**Example** <br>
<br>
Create list of words in `txtDoc3`. <br>
Count distinct words.
</font>

In [44]:
print(txtDoc3.flatMap(lambda x: x.split()).distinct().collect())
print('Number of words is %s' % (txtDoc3.flatMap(lambda x: x.split()).distinct().count()))

['constantly', 'would', 'by', 'changes', 'Do', 'Then', 'also', 'she', 'your', 'For', 'the', 'change', 'love', 'moon', 'not', 'swear']
Number of words is 16


#### Collect

Syntax is:

`collect()`

This function is already familiar. It does not restrict the output and may potentially cause out-of-memory errors of Driver.
Used with small RDDs

#### Take()

Syntax is:

`.take(n)`

Returns the first `n` elements.
This function is not deterministic. 
If run on a distributed RDD several times the results may differ.

#### Top()

Syntax is:

`.top(n, key=None)`

Returns top `n` elements of the RDD, ordered and in descending order.
This function is defined by the object type: numerical order for integers, dictionary order for strings, etc.
Argument `key` specifies the key by which to order.


<font color=red>**Exercise. Select First and Last Words in Alphabetic Order**</font> <br>
<font color=blue>
<br>
1 Print list of lower case separated words of `txtDoc3` in alphabetic order <br>
2 Print first 3 words <br>
3 Print last 3 words <br>
<br>
['also', 'by', 'change'] <br>
['your', 'would', 'then'] <br>
</font>

In [45]:
#Skipped code
txtDoc3lower=txtDoc3b.flatMap(lambda x: x.split(" ")).map(lambda x: x.lower()).sortBy(lambda x: x[0], ascending=True)
txtDoc3lower.collect()
txtDoc3lower.take(3)
txtDoc3lower.top(3)


['your', 'would', 'then']

#### First()

Syntax is:

`.first()`

Returns the first element of the RDD. <br>
This function is not deterministic. <br>
Difference from `.take(1)`:  `take(1)` returns a list, but `.first()` returns an atomic variable.

#### Reduce()

Syntax is:

`.reduce(<function>)`

This is an aggregate action. <br>
It executes a commutative and associative operations. <br>
Operation $\circ$ is commutative iff $x \circ y \equiv y \circ x$. <br>
Operation $\circ$ is associative iff $(x \circ y) \circ z \equiv x \circ (y \circ z)$. 
The specified function should have 2 inputs representing sequential values in the RDD: 

`(lambda x,y: f(x,y))`

<font color=green>
**Example** <br>
<br>
Add values of a vector. <br>
</font>

In [69]:
data = [1, 2, 3, 4, 5]
distData = sc.parallelize(data)
distData.reduce(lambda a, b: a + b) 

15

#### Fold()

Syntax is:

`.fold(zeroValue, <function>)`

This action aggregates elements of each partition of the RDD, then aggregates the results using given commutative and associative function and a `zeroValue`.
This action is similar to `.reduce()`, but can operate with empty RDDs

<font color=green>
**Example** <br>
<br>
Add values of a vector. <br>
</font>

In [46]:
numbers=sc.parallelize([1, 2, 3, 4, 5, 6, 7, 8, 9])
numbers.fold(0,lambda x, y: x+y)

45

<font color=green>
Add values of an emty RDD.
</font>

In [47]:
empty=sc.parallelize([])

<font color=green>
`reduce()` does not work: <br>
<br>
empty.reduce(lambda x, y: x+y) <br>
<br>
"ValueError: Can not reduce() empty RDD" <br>
</font>

In [48]:
empty.fold(0,lambda x, y: x+y)

0

#### TakeSample()

This is an action used to return a random list of values (elements) from the original RDD.
This function is similar to transformation `sample()`, except for form of output.
The syntax is:

`.takeSample(withReplacement, num, seed=None)`,

where `num` is the number of randomly selected recoords to be returned.

<font color=green>
**Example** <br>
<br>
Select randomly 3 out of 10 integers with fixed seed.
</font>

In [49]:
dataset=sc.parallelize([1,2,3,4,5,6,7,8,9,10])
dataset.takeSample(withReplacement=False,num=3,seed=5)

[3, 4, 2]

## Key-Value Pair Operations

### Key-Value Pair RDD Dictionary Functions

These functions return a set of keys or values from a pair RDD.

#### Keys()

Syntax is:

`.keys()`

This transformation returns an RDD with the keys from a pair RDD, or the first element from each tuple in a key-value pair RDD. 

<font color=green>
**Example** <br>
<br>

In [85]:
kvpairs=sc.parallelize([('city','Chicago'),
                        ('state','Illinois'),
                        ('zip',60637),
                        ('country','USA')])
kvpairs.keys().collect()

['city', 'state', 'zip', 'country']

#### Values()

Syntax is:

`.values()`

Returns values of each element or second element of each tuple of key-value pair RDD.


<font color=green>
**Example** <br>
<br>

In [76]:
kvpairs=sc.parallelize([('city','Chicago'),
                        ('state','Illinois'),
                        ('zip',60637),
                        ('country','USA')])
kvpairs.values().collect()

['Chicago', 'Illinois', 60637, 'USA']

### Functional Key-Value Pair RDD Transformations

These work similarly to general functional transformations. But they operate on either key or value within a tuple.

#### KeyBy()

Syntax is:

`.keyBy(<function>)`

This transformation creates key-value tuples from elements in RDD by applying function.

<font color=green>
**Example** <br>
<br>
Use location number as key in the following RDD


In [77]:
locations=sc.parallelize([('Rio de Janeiro','Brazil',1),
                         ('London','UK',2),
                          ('Beijing','China',3),
                          ('Athens','Greece',4)
                         ])
bylocno=locations.keyBy(lambda x: x[2])
bylocno.collect()

[(1, ('Rio de Janeiro', 'Brazil', 1)),
 (2, ('London', 'UK', 2)),
 (3, ('Beijing', 'China', 3)),
 (4, ('Athens', 'Greece', 4))]

#### MapValues()

Syntax is:

`.mapValues(<function>)`

Applies function to value of each key-value pair leaving the key unchanged.

#### FlatMapValues()

Syntax is:

`.flatMapValues(<function>)`

Like in general case, works like `mapValues()`, but produces a flattened list.

<font color=green>
**Example** <br>
<br>
Consider last 4 Olympic Games cities followed by numbers of nations, competitors, represented sports, disciplines and events separated by pipes.

In [51]:
OlympStats=sc.parallelize(['Rio de Janeiro,207|11303|28|41|306',
                         'London,204|10768|26|39|302',
                          'Beijing,204|10942|28|41|302',
                          ('Athens,201|10625|28|40|301')
                         ])
kvpairs=OlympStats.map(lambda x: x.split(','))
kvpairs.collect()

[['Rio de Janeiro', '207|11303|28|41|306'],
 ['London', '204|10768|26|39|302'],
 ['Beijing', '204|10942|28|41|302'],
 ['Athens', '201|10625|28|40|301']]

<font color=green>

Create list of city-value tuples.

In [54]:
locWithStats=kvpairs.flatMapValues(lambda x: x.split('|')) \
.map(lambda x: (x[0],int(x[1]))) 
locWithStats.collect()


[('Rio de Janeiro', 207),
 ('Rio de Janeiro', 11303),
 ('Rio de Janeiro', 28),
 ('Rio de Janeiro', 41),
 ('Rio de Janeiro', 306),
 ('London', 204),
 ('London', 10768),
 ('London', 26),
 ('London', 39),
 ('London', 302),
 ('Beijing', 204),
 ('Beijing', 10942),
 ('Beijing', 28),
 ('Beijing', 41),
 ('Beijing', 302),
 ('Athens', 201),
 ('Athens', 10625),
 ('Athens', 28),
 ('Athens', 40),
 ('Athens', 301)]

<font color=green>
Create list of tuples of city names and list of values.

In [87]:
locWithStatsList=kvpairs.mapValues(lambda x: x.split('|')) \
.mapValues(lambda x: [int(s) for s in x])                  
locWithStatsList.collect()

[('Rio de Janeiro', [207, 11303, 28, 41, 306]),
 ('London', [204, 10768, 26, 39, 302]),
 ('Beijing', [204, 10942, 28, 41, 302]),
 ('Athens', [201, 10625, 28, 40, 301])]

<font color=green> 
Note that `mapValues()` creates one element per city containing the city name and a list of statistics. <br>
But, `flatMapValues()` creates flattened list with 5 elements per city with city name and one of the statistics.

### Grouping, Aggregation, Sorting and Set Operations

These operations work like general grouping-aggregation-sorting functions, but specialized for key-value pairs RDD.

**Operations from this group may require repartitioning or a shuffle, causing reallocation of data between executors**.

#### GroupByKey()

Syntax is:

`.groupByKey(numPartitions=None, partitionFunc=<hash_fn>)`

This transformation groups the values for each key in a key-value pair RDD into a single sequence.
Argument `numPartitions` tells how many partitions (groups) to create.
Partitions are created by `partitionFunc` function.

Apply `groupByKey()` to Olympic locations with statistics of the previous example in order to calculate average value.

<font color=green>
**Example** <br>
<br>
Group olympic cities RDD by key

In [55]:
groupedStats=locWithStats.groupByKey()
groupedStats.collect()

[('Athens', <pyspark.resultiterable.ResultIterable at 0x1e64444aeb8>),
 ('Rio de Janeiro', <pyspark.resultiterable.ResultIterable at 0x1e64444af60>),
 ('Beijing', <pyspark.resultiterable.ResultIterable at 0x1e64444a2b0>),
 ('London', <pyspark.resultiterable.ResultIterable at 0x1e64444a668>)]

In [56]:
groupedStats.mapValues(lambda x: sum(x)/len(x)).collect() 

[('Athens', 2239.0),
 ('Rio de Janeiro', 2377.0),
 ('Beijing', 2303.4),
 ('London', 2267.8)]

<font color=green>
Note that `.groupByKey()` returns `resultIterable` type for grouped objects. <br>
In Python iterable is a sequence object that can be looped over.

**Consider using `reduceByKey()` or `foldByKey()` instead of `groupByKey()` for the purpose of aggregation (i.e. sum or count per key). Reduce and fold functions aggregate before suffling.**

#### ReduceByKey()

Syntax is:

`.reduceByKey(<function>, numPartitions=None, partitionFunc=<hashfunc>)`

This transformation merges the values for each key using an associative function. It is called on a set of key-value pairs and returns a set of key-value pairs, aggregating values for each key. Reduce function is of type 
$$v_1, v_2 %=>% v_{result}.$$

Argument `numPartitions` is effectively number of reduce tasks to execute. It can be increased to increase level of parallelization. It also affects the number of files produced by `saveAsTextFile` or other file producing actions.

Example below averages values like the previous example shows.

<font color=red>**Exercise. Average values**</font> <br>
<font color=blue>
<br>
1 Create key-value pair RDD with city name as key and tuple value statistic and combinator index:  <br>
<br>
[('Rio de Janeiro', (207, 1)), <br>
 ('Rio de Janeiro', (11303, 1)), <br>
 ('Rio de Janeiro', (28, 1)), <br>
 ('Rio de Janeiro', (41, 1)), <br>
 ('Rio de Janeiro', (306, 1)), <br>
 ('London', (204, 1)), <br>
 ('London', (10768, 1)),... <br>
 <br>
 2 Add each values of tuples by city name: <br>
 <br>
 [('Rio de Janeiro', (11885, 5)), <br>
 ('Beijing', (11517, 5)), <br>
 ('Athens', (11195, 5)), <br>
 ('London', (11339, 5))] <br>
 <br>
 3 Calculate average per city:  <br>
 <br>
 [('Rio de Janeiro', 2377.0), <br>
 ('Beijing', 2303.4), <br>
 ('Athens', 2239.0), <br>
 ('London', 2267.8)] <br>
 <br>
 Insert code in the following 3 cells. <br>

In [57]:
#Skipped code
#Create key-value pair RDD with city name as key and tuple value statistic and combinator index: 
#locWithStats.collect()
loc1=locWithStats.keyBy(lambda x: x[0]).mapValues(lambda x:(int(x[1]),int(1)))
loc1.collect()



[('Rio de Janeiro', (207, 1)),
 ('Rio de Janeiro', (11303, 1)),
 ('Rio de Janeiro', (28, 1)),
 ('Rio de Janeiro', (41, 1)),
 ('Rio de Janeiro', (306, 1)),
 ('London', (204, 1)),
 ('London', (10768, 1)),
 ('London', (26, 1)),
 ('London', (39, 1)),
 ('London', (302, 1)),
 ('Beijing', (204, 1)),
 ('Beijing', (10942, 1)),
 ('Beijing', (28, 1)),
 ('Beijing', (41, 1)),
 ('Beijing', (302, 1)),
 ('Athens', (201, 1)),
 ('Athens', (10625, 1)),
 ('Athens', (28, 1)),
 ('Athens', (40, 1)),
 ('Athens', (301, 1))]

In [58]:
#Skipped code
#loc2=loc1.groupByKey().mapValues(lambda x: sum(x))
loc2=loc1.reduceByKey(lambda x,y : (x[0]+y[0],x[1]+y[1]))
loc2.collect()

[('Athens', (11195, 5)),
 ('Rio de Janeiro', (11885, 5)),
 ('Beijing', (11517, 5)),
 ('London', (11339, 5))]

In [59]:
#Skipped code
avg=loc2.mapValues(lambda x: x[0]/x[1])
avg.collect()


[('Athens', 2239.0),
 ('Rio de Janeiro', 2377.0),
 ('Beijing', 2303.4),
 ('London', 2267.8)]

<font color=blue>
In this example average could not be done directly because it is not an associative operation.
That is why we first created tuples of totals and counts using combiner by key (both are associative and commutative operations), then we computed average as a final step.

Using combiner is a local operation done before shuffling. Then reduce step sums list of local sums instead of summing a bigger list. Thus saving time.

#### FoldByKey()

Syntax is `.foldByKey(zeroValue, <function>, numPartitions=None, partitionFunc=<hash_fn)`

This transformation is similar to `fols()` action, but specialized for key-value pairs elements. `ZeroValue` again is used to handle empty RDDs.

The function argument is of the same type 
$$v_1, v_2 %=>% v_{result}$$
as for `reduceByKey`.

<font color=green>
**Example** <br>
<br>
The following example looks for maximum by key. <br>
Obviously, the maximum number will correspond to the number of athletes in each Olympic Games.

In [60]:
maxbycity=locWithStats.foldByKey(0,lambda x,y: x if x > y else y)
maxbycity.collect()

[('Athens', 10625),
 ('Rio de Janeiro', 11303),
 ('Beijing', 10942),
 ('London', 10768)]

#### SortByKey()

Syntax is:

`.sortByKey(asscending=True, numPartitions=None, keyfunc=<function>)`

This transformation sorts a key-value pair RDD by the predefined key.
Difference from `sort()`: this time there is no need for a functions identifying the key.

<font color=red>**Exercise. Sorting RDDs**</font> <br>
<br>
<font color=blue>
In the following example with Olympic locations `locWithStats` from previous examples: <br>
<br>
1 First, sort by city: <br>
<br>
[('Athens', 201), <br>
 ('Athens', 10625), <br>
 ('Athens', 28), <br>
 ('Athens', 40), <br>
 ('Athens', 301), <br>
 ('Beijing', 204), <br>
 ('Beijing', 10942), <br>
 ('Beijing', 28), <br>
 ('Beijing', 41), <br>
 <br>
 2 Then sort by value <br>
 <br>
 [(11303, 'Rio de Janeiro'), <br>
 (10942, 'Beijing'), <br>
 (10768, 'London'), <br>
 (10625, 'Athens'), <br>
 (306, 'Rio de Janeiro'), <br>
 (302, 'London'), <br>
 (302, 'Beijing'), <br>
 (301, 'Athens'), <br>
 (207, 'Rio de Janeiro') <br>
 <br>
 Enter code in the following 2 cells. <br>

In [61]:
#Skipped code
locCitySort = locWithStats.sortByKey()
locCitySort.collect()

[('Athens', 201),
 ('Athens', 10625),
 ('Athens', 28),
 ('Athens', 40),
 ('Athens', 301),
 ('Beijing', 204),
 ('Beijing', 10942),
 ('Beijing', 28),
 ('Beijing', 41),
 ('Beijing', 302),
 ('London', 204),
 ('London', 10768),
 ('London', 26),
 ('London', 39),
 ('London', 302),
 ('Rio de Janeiro', 207),
 ('Rio de Janeiro', 11303),
 ('Rio de Janeiro', 28),
 ('Rio de Janeiro', 41),
 ('Rio de Janeiro', 306)]

In [62]:
#Skipped code
locWithStats.sortBy(lambda x : x[1], ascending = False).map(lambda x: (x[1],x[0])).collect()

[(11303, 'Rio de Janeiro'),
 (10942, 'Beijing'),
 (10768, 'London'),
 (10625, 'Athens'),
 (306, 'Rio de Janeiro'),
 (302, 'London'),
 (302, 'Beijing'),
 (301, 'Athens'),
 (207, 'Rio de Janeiro'),
 (204, 'London'),
 (204, 'Beijing'),
 (201, 'Athens'),
 (41, 'Rio de Janeiro'),
 (41, 'Beijing'),
 (40, 'Athens'),
 (39, 'London'),
 (28, 'Rio de Janeiro'),
 (28, 'Beijing'),
 (28, 'Athens'),
 (26, 'London')]

<font color=blue>
Again, the largest values on top should be the records of numbers of atheletes in each Olympics.

#### SubtractByKey()

Syntax is:

`RDD1.subtractByKey(RDD2,numPartitions=None)`

This is a set operation similar to `subtract()` transformation, it returns key-value pairs from `RDD1` with keys not present in `RDD2`.

<font color=green>
**Example** <br>
<br>
In the following example consider two RDDs: last 4 summer and winter Olympic Games hosting continents and tuples of number of nations and number of events.

In [63]:
locSummerOlymp=sc.parallelize([('America',(207,306)),
                         ('Europe',(204,302)),
                          ('Asia',(204,302)),
                          ('Europe',(201,301))
                         ])
locWinterOlymp=sc.parallelize([('Europe',(88,98)),
                         ('America',(82,86)),
                          ('Europe',(80,84)),
                          ('America',(78,78))
                         ])
print(locSummerOlymp.subtractByKey(locWinterOlymp).collect())
print(locWinterOlymp.subtractByKey(locSummerOlymp).collect())

[('Asia', (204, 302))]
[]


### Join Functions

Join functions combine values from 2 RDDs on a common key field.
The 2 datasets are referred to in the order they are specified: the first specified dataset is considered **left**, the second is considered **right**.

Types of joins:

-  Inner join, or simply join, returnes all elements or records from both datasets where the key is present in both datasets;
-  Outer join does not require keys to match in both datasets. Outer join can be left outer, right outer of full outer join;
-  Left outer join returns all records from the left along with matched by key records from the right;
-  Right outer join returns all records from the right along with matched by key records from the left;
-  Full outer join returns all records from both datasets regardless of whether keys match or not.

#### Join()

Syntax is:

`RDD1.join(RDD2, numpartitions=None)`

This transformation implements inner join. Argument `numPartitions` determines how many partitions to create of the output. <br>
Returned RDD containes a tuple with matched key and a tuple of two complete records with matched key, each in the form of a list. 

<font color=red>**Exercise. Joining RDDs**</font> <br>
<font color=blue>
<br>
Create the following RDDs: 

*  `stores` - with stores and store locations; <br>
*  `salespeople` - with names of salespeople and stores they are assigned to. <br>
<br>
1 Prepare the data for analysis:  <br>  
<br>
*  Split records by `"\t"` <br>
*  Make keys: first part of element of `stores` and last part of element of `salespeople` <br>
<br>
Resulting RDDs should look like: <br>
<br>
-  `storestr` <br>
<br>
[('100', ['100', 'Boca Raton']), <br>
 ('101', ['101', 'Columbia']), <br>
 ('102', ['102', 'Cambridge']), <br>
 ('103', ['103', 'Naperville'])] <br>
 <br>
-  `salespeopletr` <br>
<br>
[('100', ['1', 'Henry', '100']), <br>
 ('100', ['2', 'Karen', '100']), <br>
 ('101', ['3', 'Paul', '101']), <br>
 ('102', ['4', 'Jimmy', '102']), <br>
 ('', ['5', 'Janice', ''])] <br>
 <br>
 2 Join `salespeopletr` (left) and `storestr` (right): <br>
 <br>
 [('102', (['4', 'Jimmy', '102'], ['102', 'Cambridge'])), <br>
 ('100', (['1', 'Henry', '100'], ['100', 'Boca Raton'])), <br>
 ('100', (['2', 'Karen', '100'], ['100', 'Boca Raton'])), <br>
 ('101', (['3', 'Paul', '101'], ['101', 'Columbia']))] <br>
 <br>
  Enter code in the cells below and show result of each of them. 

In [64]:
stores=sc.parallelize(['100\tBoca Raton','101\tColumbia','102\tCambridge','103\tNaperville']) 
stores.collect()

['100\tBoca Raton', '101\tColumbia', '102\tCambridge', '103\tNaperville']

In [65]:
salespeople=sc.parallelize(['1\tHenry\t100','2\tKaren\t100', \
                            '3\tPaul\t101','4\tJimmy\t102','5\tJanice\t'])
salespeople.collect()

['1\tHenry\t100',
 '2\tKaren\t100',
 '3\tPaul\t101',
 '4\tJimmy\t102',
 '5\tJanice\t']

In [66]:
#Skipped code
storestr = stores.map(lambda x: x.split("\t")).keyBy(lambda x: x[0])
storestr.collect()

[('100', ['100', 'Boca Raton']),
 ('101', ['101', 'Columbia']),
 ('102', ['102', 'Cambridge']),
 ('103', ['103', 'Naperville'])]

In [67]:
#Skipped code
#salespeopletr
salespeopletr = salespeople.map(lambda x: x.split("\t")).keyBy(lambda x: x[2])
salespeopletr.collect()

[('100', ['1', 'Henry', '100']),
 ('100', ['2', 'Karen', '100']),
 ('101', ['3', 'Paul', '101']),
 ('102', ['4', 'Jimmy', '102']),
 ('', ['5', 'Janice', ''])]

In [68]:
#Skipped code
salesstore=salespeopletr.join(storestr)
salesstore.collect()

[('102', (['4', 'Jimmy', '102'], ['102', 'Cambridge'])),
 ('101', (['3', 'Paul', '101'], ['101', 'Columbia'])),
 ('100', (['1', 'Henry', '100'], ['100', 'Boca Raton'])),
 ('100', (['2', 'Karen', '100'], ['100', 'Boca Raton']))]

**Think about optimal use of join operations: "join large by small", that is: use larger RDD as left and smaller as right**.

#### LeftOuterJoin()

Syntax is:

`RDD1.leftOuterJoin(RDD2, numpertitions=None)`

This transformation returns all records of the left RDD. If a key from the left is also present in the right RDD then the record from the right will be returned along with the record from the left for the same key.

<font color=blue>

Apply left outer join to the transformed RDDs: <br>
<br>
[('102', (['4', 'Jimmy', '102'], ['102', 'Cambridge'])), <br>
 ('', (['5', 'Janice', ''], None)), <br>
 ('100', (['1', 'Henry', '100'], ['100', 'Boca Raton'])), <br>
 ('100', (['2', 'Karen', '100'], ['100', 'Boca Raton'])), <br>
 ('101', (['3', 'Paul', '101'], ['101', 'Columbia']))] <br>

In [69]:
#Skipped code
salesstoreLOJ=salespeopletr.leftOuterJoin(storestr)
salesstoreLOJ.collect()

[('102', (['4', 'Jimmy', '102'], ['102', 'Cambridge'])),
 ('', (['5', 'Janice', ''], None)),
 ('101', (['3', 'Paul', '101'], ['101', 'Columbia'])),
 ('100', (['1', 'Henry', '100'], ['100', 'Boca Raton'])),
 ('100', (['2', 'Karen', '100'], ['100', 'Boca Raton']))]

<font color=blue>
Use the joint RDD to find salespeople not assigned to stores: <br>
<br>
1 `result1`: Select records for which right RDD (`storestr`) does not have a key from the left (`storestr`), i.e. salesperson without a store association; <br>
2 `result2`: Return a list of text messages for each salesperson like this: "salesperson Name has no store" <br>
<br>
 Enter code in the following 2 cells and show result of each of them. <br>
 <br>
 [('', (['5', 'Janice', ''], None))] <br>
 <br>
 ['salesperson Janice has no store']

In [73]:
#Skipped code
spwithoutstore=salesstoreLOJ.subtractByKey(salesstore)
spwithoutstore.collect()


[('', (['5', 'Janice', ''], None))]

In [74]:
#Skipped code

print(['salesperson Janice has no store'])



['salesperson Janice has no store']


#### RightOuterJoin()

Syntax is:

`RDD1.rightOuterJoin(RDD2, numPartitions)`

This transformation returns all records of the right RDD. If a key from the right is also present in the left RDD then the record from the left will be returned along with the record from the right for the same key.

<font color=blue>

Use this transformation to find stores to which there are no assigned salespeople. <br>
<br>
['Naperville store has no salespeople']

In [77]:
#Skipped code
nosalesROJ=salespeopletr.rightOuterJoin(storestr)
nosalesROJ.collect()
print(['Naperville store has no salespeople'])

[('102', (['4', 'Jimmy', '102'], ['102', 'Cambridge'])),
 ('101', (['3', 'Paul', '101'], ['101', 'Columbia'])),
 ('100', (['1', 'Henry', '100'], ['100', 'Boca Raton'])),
 ('100', (['2', 'Karen', '100'], ['100', 'Boca Raton'])),
 ('103', (None, ['103', 'Naperville']))]

#### FullOuterJoin()

Syntax is:

`RDD1.fullOuterJoin(RDD2, numPartitions=None)`

This transformation returns an RDD of all elements from both left and right source RDDs. Key not matched are represented by None or empty records for the corresponding left or right RDD.

<font color=blue>
**Assignment** <br>
<br>
Combine stores and salespeople RDDs using full outer join.

[('', (['5', 'Janice', ''], None)), ('103', (None, ['103', 'Naperville']))]

In [78]:
#Skipped code
salesstoreROJ = salespeopletr.fullOuterJoin(storestr)
salesstoreROJ.collect()

[('102', (['4', 'Jimmy', '102'], ['102', 'Cambridge'])),
 ('', (['5', 'Janice', ''], None)),
 ('101', (['3', 'Paul', '101'], ['101', 'Columbia'])),
 ('100', (['1', 'Henry', '100'], ['100', 'Boca Raton'])),
 ('100', (['2', 'Karen', '100'], ['100', 'Boca Raton'])),
 ('103', (None, ['103', 'Naperville']))]

#### Cogroup()

Syntax is:

`RDD1.cogroup(RDD2,numPartitions=None)`

This transformation groups multiple key-value pair datasets by a key. <br>
Differences from `fullOuterJoin()`:

1 Transformation `cogroup()` returns iterable object, similar to `groupByKey()` <br>
2 Transformation `cogroup()` groups multiple elements from both RDDs into iterable objects, whereas `fullOuterJoin()` creates separate lists for the same key
3 Transformation `cogroup()` can group 3 or more RDDs

The result of `cogroup()` of 2 RDDs `A` and `B`  with a key `K` looks like 

$$[K, Iterable(K, V_A, \ldots), Iterable (K, V_B, \ldots)]$$

If one of the source RDDs  does not have have elements with the same key as another source RDD then `cogroup()` returns empty Iterable.

<font color=blue>
**Assignment** <br>

Check that <br>

`salespeopletr.cogroup(storestr).collect()`

shows iterable results.
   
Add code in the following cell

In [79]:
#Skipped code
salespeopletr.cogroup(storestr).collect()


[('102',
  (<pyspark.resultiterable.ResultIterable at 0x1e644473c88>,
   <pyspark.resultiterable.ResultIterable at 0x1e6444733c8>)),
 ('',
  (<pyspark.resultiterable.ResultIterable at 0x1e6444737f0>,
   <pyspark.resultiterable.ResultIterable at 0x1e644465710>)),
 ('101',
  (<pyspark.resultiterable.ResultIterable at 0x1e644465240>,
   <pyspark.resultiterable.ResultIterable at 0x1e644465d30>)),
 ('100',
  (<pyspark.resultiterable.ResultIterable at 0x1e6444655c0>,
   <pyspark.resultiterable.ResultIterable at 0x1e644465c50>)),
 ('103',
  (<pyspark.resultiterable.ResultIterable at 0x1e644465400>,
   <pyspark.resultiterable.ResultIterable at 0x1e644454d68>))]

<font color=blue>
Apply `cogroup()` as:

salespeopletr.cogroup(storestr) \
.mapValues(lambda x: [item for sublist in x for item in sublist]) \
.collect()

Explain the result.

In [82]:
#Skipped code
salespeopletr.cogroup(storestr) \
.mapValues(lambda x: [item for sublist in x for item in sublist]) \
.collect()
# Here we see store as a key, the value is a list of 2 lists. 
#elements are the values from each of the datasets, if available in both

[('102', [['4', 'Jimmy', '102'], ['102', 'Cambridge']]),
 ('', [['5', 'Janice', '']]),
 ('101', [['3', 'Paul', '101'], ['101', 'Columbia']]),
 ('100',
  [['1', 'Henry', '100'], ['2', 'Karen', '100'], ['100', 'Boca Raton']]),
 ('103', [['103', 'Naperville']])]

#### Cartesian()

Syntax is:

`RDD1.cartesian(RDD2)`

This transformation (called also **cross join**) generates every possible combination of records from two RDDs with the number of records equal to the product of numbers of records of the source RDDs.

<font color=blue>
**Assignment** <br>
Apply to `salespeopletr` on the left and `storestr` on the right.

[(('100', ['1', 'Henry', '100']), ('100', ['100', 'Boca Raton'])), <br>
 (('100', ['1', 'Henry', '100']), ('101', ['101', 'Columbia'])), <br>
 (('100', ['1', 'Henry', '100']), ('102', ['102', 'Cambridge'])), <br>
 (('100', ['1', 'Henry', '100']), ('103', ['103', 'Naperville'])), <br>
 (('100', ['2', 'Karen', '100']), ('100', ['100', 'Boca Raton'])), <br>
 (('100', ['2', 'Karen', '100']), ('101', ['101', 'Columbia'])), <br>
 (('100', ['2', 'Karen', '100']), ('102', ['102', 'Cambridge'])), <br>
 (('100', ['2', 'Karen', '100']), ('103', ['103', 'Naperville'])), <br>
 (('101', ['3', 'Paul', '101']), ('100', ['100', 'Boca Raton'])), <br>
 (('101', ['3', 'Paul', '101']), ('101', ['101', 'Columbia'])), <br>
 (('101', ['3', 'Paul', '101']), ('102', ['102', 'Cambridge'])), <br>
 (('101', ['3', 'Paul', '101']), ('103', ['103', 'Naperville'])), <br>
 (('102', ['4', 'Jimmy', '102']), ('100', ['100', 'Boca Raton'])), <br>
 (('', ['5', 'Janice', '']), ('100', ['100', 'Boca Raton'])), <br>
 (('102', ['4', 'Jimmy', '102']), ('101', ['101', 'Columbia'])), <br>
 (('', ['5', 'Janice', '']), ('101', ['101', 'Columbia'])), <br>
 (('102', ['4', 'Jimmy', '102']), ('102', ['102', 'Cambridge'])), <br>
 (('', ['5', 'Janice', '']), ('102', ['102', 'Cambridge'])), <br>
 (('102', ['4', 'Jimmy', '102']), ('103', ['103', 'Naperville'])), <br>
 (('', ['5', 'Janice', '']), ('103', ['103', 'Naperville']))] <br>

In [83]:
#Skipped code
cart=salespeopletr.cartesian(storestr)
cart.collect()

[(('100', ['1', 'Henry', '100']), ('100', ['100', 'Boca Raton'])),
 (('100', ['1', 'Henry', '100']), ('101', ['101', 'Columbia'])),
 (('100', ['1', 'Henry', '100']), ('102', ['102', 'Cambridge'])),
 (('100', ['1', 'Henry', '100']), ('103', ['103', 'Naperville'])),
 (('100', ['2', 'Karen', '100']), ('100', ['100', 'Boca Raton'])),
 (('100', ['2', 'Karen', '100']), ('101', ['101', 'Columbia'])),
 (('100', ['2', 'Karen', '100']), ('102', ['102', 'Cambridge'])),
 (('100', ['2', 'Karen', '100']), ('103', ['103', 'Naperville'])),
 (('101', ['3', 'Paul', '101']), ('100', ['100', 'Boca Raton'])),
 (('101', ['3', 'Paul', '101']), ('101', ['101', 'Columbia'])),
 (('101', ['3', 'Paul', '101']), ('102', ['102', 'Cambridge'])),
 (('101', ['3', 'Paul', '101']), ('103', ['103', 'Naperville'])),
 (('102', ['4', 'Jimmy', '102']), ('100', ['100', 'Boca Raton'])),
 (('102', ['4', 'Jimmy', '102']), ('101', ['101', 'Columbia'])),
 (('102', ['4', 'Jimmy', '102']), ('102', ['102', 'Cambridge'])),
 (('102', [

**Be careful using this transformation because it may create a very large RDD**.